File: crates/core/src/agent.rs
use std::fmt;
use std::sync::Arc;

use async_trait::async_trait;

use crate::context::Context;
// ============================================================================
// 6. Agent loop seam
// ============================================================================

#[derive(Debug, Clone)]
pub struct AgentConfig {
    pub model: String,
    pub system_prompt: String,
    pub max_steps: usize,
    pub max_parallel_tool_calls: usize,
}

impl Default for AgentConfig {
    fn default() -> Self {
        Self {
            model: "deepseek-chat".into(),
            system_prompt:
                "You are celestea, an AI agent. You are concise, accurate and direct."
                    .into(),
            max_steps: 16,
            max_parallel_tool_calls: 4,
        }
    }
}

#[derive(Debug, Clone)]
pub struct AgentError(pub String);

impl fmt::Display for AgentError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for AgentError {}

#[async_trait]
pub trait AgentLoop: Send + Sync {
    /// Drive one user turn: append to the session log, loop over model steps,
    /// dispatch tool calls, and write the final assistant message.
    async fn run_turn(&self, ctx: &Context, user_input: &str) -> Result<(), AgentError>;
}

pub struct AgentLoopService(pub Arc<dyn AgentLoop>);
impl std::ops::Deref for AgentLoopService {
    type Target = dyn AgentLoop;
    fn deref(&self) -> &Self::Target {
        &*self.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn agent_config_default_has_explicit_identity() {
        let cfg = AgentConfig::default();
        // C·identity: the default system prompt states an explicit agent identity
        // instead of the old generic "helpful assistant" boilerplate.
        assert!(
            cfg.system_prompt.contains("celestea"),
            "default system_prompt should name the agent identity, got: {}",
            cfg.system_prompt
        );
        assert!(cfg.system_prompt.contains("concise"));
        assert!(!cfg.system_prompt.contains("helpful assistant"));
    }

}

File: packages/core/src/agent.ts
/**
 * AgentLoop seam — port of `crates/core/src/agent.rs`.
 *
 * The loop drives one user turn against a Context: append to the session log,
 * step the model, dispatch tool calls, write the final assistant message. The
 * concrete loop lives in packages/agent-loop and is plugged in at compose time.
 */

import type { Context } from "./context.js";

export interface AgentConfig {
  model: string;
  system_prompt: string;
  max_steps: number;
  max_parallel_tool_calls: number;
  /** Model context window in tokens; 0 disables context trimming. */
  context_window_tokens: number;
  /** Trim factor (0..=1) of the window that triggers old-message trimming. */
  context_trim_threshold: number;
  /** How many most-recent messages to keep when trimming (plus system). */
  context_keep_recent: number;
}

/** Rust `AgentConfig::default()` — values are contract (identity prompt included). */
export function defaultAgentConfig(): AgentConfig {
  return {
    model: "deepseek-chat",
    system_prompt: "You are celestea, an AI agent. You are concise, accurate and direct.",
    max_steps: 16,
    max_parallel_tool_calls: 4,
    context_window_tokens: 65_536,
    context_trim_threshold: 0.8,
    context_keep_recent: 10,
  };
}

export class AgentError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AgentError";
  }
}

export interface AgentLoop {
  /** Drive one user turn; rejects with [AgentError] on a terminal failure. */
  runTurn(ctx: Context, userInput: string): Promise<void>;
}

/** Well-known token for the agent loop service in a Context. */
export const AGENT_LOOP_SERVICE = "celestea.core.AgentLoop";
